control/trajectory_follower/src/pid_longitudinal_controller.cpp

```diff
@@ -373,36 +377,40 @@ boost::optional<LongitudinalOutput> PidLongitudinalController::run()
   // calculate current pose and control data
   geometry_msgs::msg::Pose current_pose = m_current_kinematic_state_ptr->pose.pose;
 
-  const auto control_data = getControlData(current_pose);
+  m_control_data = getControlData(current_pose);
```

Comment:
First of all, control data is required data to calculate PID control. (This should be explicitly commented though)
Therefore, I guess we should not put trans_dev, and rot_dev in control_data.

Second, I created control_data as a local variable in order to make member variable less.
If control_data is member variable, which means a kind of global variable, it can be modified from anywhere. To make maintainability higher, I feel the local variable is better in this case.
What do you think?

Reply:
> First of all, control data is required data to calculate PID control. (This should be explicitly commented though)
Therefore, I guess we should not put trans_dev, and rot_dev in control_data.

This diagnostic needs trans_dev and rot_dev to confirm how to decide as emergency.
I talk with @TakaHoribe to realize it, then he said you can make control_data as member variable and add trans_dev and rot_dev because they are calculated in getControlData().

> If control_data is member variable, which means a kind of global variable, it can be modified from anywhere. To make maintainability higher, I feel the local variable is better in this case.

I agree with what you say.
instead of it, Can I make rot_dev and trans_dev as member variables ?

Reply:
Sorry to confuse you. I got it.

This is my opinion.
The reason we have to have some member variables is due to diagnostic updater callback.
To explicitly express it, I would create diagnostic_data (DiagnosticData struct) as a member variable.
Then update it in getControlData with is_dist_deviation_large and is_yaw_deviation_large.

If we adopted this idea, DiagnosticData would be
```cpp
struct DiagnosticData
{
  is_dist_deviation_large;
  is_yaw_deviation_large;
};
```

Reply:
Thank you for your opinion.
But, I want calculated value to compare it to its threshold, 
In my opinion, DiagnosticsData woud be

``` cpp
struct DiagnosticData
{
    float64_t trans_deviation{0.0};  // translation deviation between nearest point and current_pose
    float64_t rot_deviation{0.0}; // rotation deviation between nearest point and current_pose
};

Reply:
Okay, looks good to me!

Reply:
I fixed it